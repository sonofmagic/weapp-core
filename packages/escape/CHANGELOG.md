# @weapp-core/escape

## 3.0.0

### Major Changes

- [`b9e3e2c`](https://github.com/sonofmagic/weapp-core/commit/b9e3e2c47c046bad901baaa32825e8e849225a3f) Thanks [@sonofmagic](https://github.com/sonofmagic)! - feat: unique separator
