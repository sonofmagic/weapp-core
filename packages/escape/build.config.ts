import path from 'node:path'
import { defineBuildConfig } from 'unbuild'

export default defineBuildConfig({
  // 可按需指定入口，例如 ['./src/index', './src/cli']
  // 默认仅构建入口示例 ['./src/index']
  rollup: {
    // 内联，相当于 nodeResolve
    inlineDependencies: true,
    // cjs
    emitCJS: true,
    // 添加 cjs 注入
    cjsBridge: true,
    dts: {
      // 相关问题记录：https://github.com/unjs/unbuild/issues/135
      respectExternal: false,
    },
  },
  alias: {
    // 别名
    '@': path.resolve(__dirname, './src'),
  },
  // dts
  declaration: true,
})
