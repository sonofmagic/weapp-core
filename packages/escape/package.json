{
  "name": "@weapp-core/escape",
  "type": "commonjs",
  "version": "2.1.0",
  "description": "escape class name in weapp",
  "author": "SonOfMagic <qq1324318532@gmail.com>",
  "license": "MIT",
  "homepage": "https://github.com/sonofmagic/weapp-core/tree/main/packages/escape#readme",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/sonofmagic/weapp-core.git"
  },
  "bugs": {
    "url": "https://github.com/sonofmagic/weapp-core/issues"
  },
  "keywords": [
    "weapp",
    "escape",
    "css"
  ],
  "sideEffects": false,
  "exports": {
    ".": {
      "types": "./dist/index.d.ts",
      "import": "./dist/index.mjs",
      "require": "./dist/index.cjs"
    }
  },
  "main": "./dist/index.cjs",
  "module": "./dist/index.mjs",
  "types": "./dist/index.d.ts",
  "files": [
    "dist"
  ],
  "scripts": {
    "build": "unbuild",
    "test": "vitest run --coverage.enabled",
    "test:dev": "vitest --coverage.enabled",
    "release": "pnpm publish"
  },
  "publishConfig": {
    "access": "public",
    "registry": "https://registry.npmjs.org"
  },
  "dependencies": {}
}
