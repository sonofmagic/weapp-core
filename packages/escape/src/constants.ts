export const SYMBOL_TABLE = {
  BACKQUOTE: '`',
  TILDE: '~',
  EXCLAM: '!',
  AT: '@',
  NUMBERSIGN: '#',
  DOLLAR: '$',
  PERCENT: '%',
  CARET: '^',
  AMPERSAND: '&',
  ASTERISK: '*',
  PARENLEFT: '(',
  PARENRIGHT: ')',
  MINUS: '-',
  UNDERSCORE: '_',
  EQUAL: '=',
  PLUS: '+',
  BRACKETLEFT: '[',
  BRACELEFT: '{',
  BRACKETRIGHT: ']',
  BRACERIGHT: '}',
  SEMICOLON: ';',
  COLON: ':',
  QUOTE: '\'',
  DOUBLEQUOTE: '"',
  BACKSLASH: '\\',
  BAR: '|',
  COMMA: ',',
  LESS: '<',
  PERIOD: '.',
  GREATER: '>',
  SLASH: '/',
  QUESTION: '?',
  SPACE: ' ',
  DOT: '.',
  HASH: '#',
} as const

export type SYMBOL_TABLE_TYPE = typeof SYMBOL_TABLE

export type SYMBOL_TABLE_TYPE_VALUES = SYMBOL_TABLE_TYPE[keyof SYMBOL_TABLE_TYPE]

export type MappingStringDictionary = Record<Exclude<SYMBOL_TABLE_TYPE_VALUES, '-' | '_' | ' '>, string>

// 需要转义的 ASCII 区间：32–47、58–64、91–96、123–126（排除 '-'、'_' 与空格）。
// 这些符号会干扰类名或 ID，统一处理可以兼顾大小写敏感的选择器。
const ESCAPE_CHARACTER_CODES = [
  ['[', 'b'],
  [']', 'B'],
  ['(', 'p'],
  [')', 'P'],
  ['#', 'h'],
  ['!', 'e'],
  ['/', 'f'],
  ['\\', 'r'],
  ['.', 'd'],
  [':', 'c'],
  ['%', 'v'],
  [',', 'm'],
  ['\'', 'a'],
  ['"', 'q'],
  ['*', 'x'],
  ['&', 'n'],
  ['@', 't'],
  ['{', 'k'],
  ['}', 'K'],
  ['+', 'u'],
  [';', 'j'],
  ['<', 'l'],
  ['~', 'w'],
  ['=', 'z'],
  ['>', 'g'],
  ['?', 'Q'],
  ['^', 'y'],
  ['`', 'i'],
  ['|', 'o'],
  ['$', 's'],
] as const satisfies ReadonlyArray<[Exclude<SYMBOL_TABLE_TYPE_VALUES, '-' | '_' | ' '>, string]>

function buildMapping(prefix: string) {
  return ESCAPE_CHARACTER_CODES.reduce<Record<string, string>>((acc, [char, code]) => {
    acc[char] = `${prefix}${code}`
    return acc
  }, {})
}

const SIMPLE_TOKEN_PREFIX = '_'
const COMPLEX_TOKEN_PREFIX = '_c'

// 复杂映射采用更长的标记，确保在外部系统中也能安全反解。
export const ComplexMappingChars2String: MappingStringDictionary = buildMapping(COMPLEX_TOKEN_PREFIX) as MappingStringDictionary

export const ComplexMappingChars2StringEntries = Object.entries(ComplexMappingChars2String)

// 简单映射追求更短的标记，适合对体积敏感的类名（如 Tailwind 用例）。
export const MappingChars2String: MappingStringDictionary = buildMapping(SIMPLE_TOKEN_PREFIX) as MappingStringDictionary

export const MappingChars2StringEntries = Object.entries(MappingChars2String)

export const MAX_ASCII_CHAR_CODE = 127
