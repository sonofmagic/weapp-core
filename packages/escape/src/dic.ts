export const SYMBOL_TABLE = {
  BACKQUOTE: '`',
  TILDE: '~',
  EXCLAM: '!',
  AT: '@',
  NUMBERSIGN: '#',
  DOLLAR: '$',
  PERCENT: '%',
  CARET: '^',
  AMPERSAND: '&',
  ASTERISK: '*',
  PARENLEFT: '(',
  PARENRIGHT: ')',
  MINUS: '-',
  UNDERSCORE: '_',
  EQUAL: '=',
  PLUS: '+',
  BRACKETLEFT: '[',
  BRACELEFT: '{',
  BRACKETRIGHT: ']',
  BRACERIGHT: '}',
  SEMICOLON: ';',
  COLON: ':',
  QUOTE: "'",
  DOUBLEQUOTE: '"',
  BACKSLASH: '\\',
  BAR: '|',
  COMMA: ',',
  LESS: '<',
  PERIOD: '.',
  GREATER: '>',
  SLASH: '/',
  QUESTION: '?',
  SPACE: ' ',
  DOT: '.',
  HASH: '#'
} as const

export type SYMBOL_TABLE_TYPE = typeof SYMBOL_TABLE

export type SYMBOL_TABLE_TYPE_VALUES = SYMBOL_TABLE_TYPE[keyof SYMBOL_TABLE_TYPE]

export type MappingStringDictionary = Record<Exclude<SYMBOL_TABLE_TYPE_VALUES, '-' | '_' | ' '>, string>

export const MappingChars2String: MappingStringDictionary = {
  '[': '_bl_',
  ']': '_br_',
  '(': '_pl_',
  ')': '_qr_',
  '#': '_h_',
  '!': '_i_',
  '/': '_s_',
  '\\': '_bs_',
  '.': '_d_',
  ':': '_c_',
  '%': '_p_',
  ',': '_co_',
  "'": '_q_',
  '"': '_dq_',
  '*': '_a_',
  '&': '_am_',
  '@': '_at_',
  '{': '_bal_',
  '}': '_bar_',
  // ' ': '_sp_',
  '+': '_plus_',
  // '-': '_m_',
  ';': '_se_',
  '<': '_l_',
  '~': '_t_',
  '=': '_e_',
  '>': '_g_',
  '?': '_qu_',
  '^': '_ca_',
  '`': '_bq_',
  '|': '_b_',
  $: '_do_'
  // _: '_u_'
} as const

export const MappingChars2StringEntries = Object.entries(MappingChars2String)
// 用 _ 比 - 好
// https://www.rapidtables.com/code/text/ascii-table.html
// ID选择器和类选择器对大小写敏感
// 标签选择器、属性选择器不区分大小写
// 同样一条 decl 你可以写 px 也可以写 PX
// 0-31 48-57(0-9数字字符) 65-90(大写英文) 97-122(小写英文) 127(DEL)不考虑
// #region 32-47 count:16
// 32	20	space // 其实可以忽略空格
// 33	21	! ✔
// 34	22	" ✔
// 35	23	# ✔
// 36	24	$ ✔
// 37	25	% ✔
// 38	26	& ✔
// 39	27	' ✔
// 40	28	( ✔
// 41	29	) ✔
// 42	2A	* ✔
// 43	2B	+ ✔
// 44	2C	, ✔
// 45	2D	- //小程序中允许，但是不能单独放在前后第一位-
// 46	2E	. ✔
// 47	2F	/ ✔
// #endregion
// #region 58-64 count:7 ✔
// 58	3A	: ✔
// 59	3B	; ✔
// 60	3C	< ✔
// 61	3D	= ✔
// 62	3E	> ✔
// 63	3F	? ✔
// 64	40	@ ✔
// #endregion
// #region 91-96 count:6 ✔
// 91	5B	[ ✔
// 92	5C	\ ✔
// 93	5D	] ✔
// 94	5E	^ ✔
// 95	5F	_ // 小程序中合法
// 96	60	` ✔
// #endregion
// #region 123-126 count:4 ✔
// 123	7B	{ ✔
// 124	7C	| ✔
// 125	7D	} ✔
// 126	7E	~ ✔
// #endregion
// 15 + 7 + 6 + 4 = 32
export const SimpleMappingChars2String: MappingStringDictionary = {
  '[': '_',
  ']': '_',
  '(': '_',
  ')': '_',
  '{': '_',
  '}': '_',
  '+': 'a',
  ',': 'b',
  ':': 'c',
  '.': 'd',
  '=': 'e',
  ';': 'f',
  '>': 'g',
  '#': 'h',
  '!': 'i',
  '@': 'j',
  '^': 'k',
  '<': 'l',
  '*': 'm',
  '&': 'n',
  '?': 'o',
  '%': 'p',
  "'": 'q',
  $: 'r',
  '/': 's',
  '~': 't',
  '|': 'u',
  '`': 'v',
  '\\': 'w',
  '"': 'x'
}

export const SimpleMappingChars2StringEntries = Object.entries(SimpleMappingChars2String)
