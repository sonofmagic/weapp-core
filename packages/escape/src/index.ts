export * from './constants'
export * from './escape'
export * from './predicates'
export * from './types'
export * from './unescape'

export const __areExportsLoaded = true
