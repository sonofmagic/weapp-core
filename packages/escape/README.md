# @weapp-core/escape

> core function to escape weapp class name

## Install

```bash
<npm | yarn | pnpm> i @weapp-core/escape
```

## Usage

```ts
import { escape } from '@weapp-core/escape'

escape(str)
```
