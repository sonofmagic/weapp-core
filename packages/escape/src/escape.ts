import type { EscapeOptions } from './types'
import { MAX_ASCII_CHAR_CODE } from './constants'
import { ESCAPE_PREFIX, HYPHEN } from './internal-constants'
import {
  createEscapeMapping,
  DEFAULT_ESCAPE_KEYS,
  DEFAULT_ESCAPE_MAPPING,
  hasOwnKey,
} from './mapping'
import { isAsciiNumber } from './predicates'

function escapeLeadingCharacter(char: string, nextChar: string | undefined, ignoreHead: boolean) {
  if (ignoreHead) {
    return char
  }

  const code = char.codePointAt(0)!

  if (isAsciiNumber(code)) {
    return `${ESCAPE_PREFIX}${char}`
  }

  if (char === HYPHEN) {
    if (!nextChar) {
      return `${ESCAPE_PREFIX}${char}`
    }
    if (isAsciiNumber(nextChar.codePointAt(0)!)) {
      return `${ESCAPE_PREFIX}${char}`
    }
  }

  return char
}

export function escape(
  selectors: string,
  options?: EscapeOptions,
) {
  if (selectors.length === 0) {
    return ''
  }

  const map = createEscapeMapping(options?.map)
  const ignoreHead = options?.ignoreHead ?? false
  const usingDefaultMap = map === DEFAULT_ESCAPE_MAPPING

  const sb: string[] = []
  const characters = Array.from(selectors)

  for (let index = 0; index < characters.length; index++) {
    const char = characters[index]
    const code = char.codePointAt(0)!

    if (code > MAX_ASCII_CHAR_CODE) {
      sb.push(`u${code.toString(16)}`)
      continue
    }

    if (usingDefaultMap) {
      if (DEFAULT_ESCAPE_KEYS.has(char)) {
        sb.push(DEFAULT_ESCAPE_MAPPING[char])
        continue
      }
    }
    else if (hasOwnKey(map, char)) {
      sb.push(map[char])
      continue
    }

    if (index === 0) {
      sb.push(escapeLeadingCharacter(char, characters[index + 1], ignoreHead))
      continue
    }

    sb.push(char)
  }

  return sb.join('')
}
