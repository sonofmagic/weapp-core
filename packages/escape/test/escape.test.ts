import { describe, expect, it } from 'vitest'

import {
  ComplexMappingChars2String as ComplexMap,
  MAX_ASCII_CHAR_CODE,
} from '../src/constants'
import { escape } from '../src/escape'

const testCase = `1234567890-=\b~!@#$%^&*()_+qwertyuiop[]\\QWERTYUIOP{}|asdfghjkl;'ASDFGHJKL:"zxcvbnm,./ZXCVBNM<>?`

describe('escape', () => {
  it('should match snapshot for default mapping', () => {
    expect(escape(testCase)).toMatchSnapshot()
  })

  it('should match snapshot for complex mapping', () => {
    expect(escape(testCase, { map: ComplexMap })).toMatchSnapshot()
  })

  it('should handle empty input', () => {
    expect(escape('')).toBe('')
  })

  it('should escape leading hyphen when needed', () => {
    expect(escape('-')).toBe('_-')
    expect(escape('-', { ignoreHead: true })).toBe('-')
  })

  it('should escape leading hyphen followed by digits', () => {
    expect(escape('-12345')).toBe('_-12345')
    expect(escape('-12345', { ignoreHead: true })).toBe('-12345')
  })

  it('should leave hyphen followed by letters untouched', () => {
    expect(escape('-plmkoi')).toBe('-plmkoi')
    expect(escape('-plmkoi', { ignoreHead: true })).toBe('-plmkoi')
  })

  it('should escape mapped characters using custom map', () => {
    expect(escape('abc', { map: { a: 'A', b: 'B' } })).toBe('ABc')
  })

  it('should return unicode escape sequences for non-ascii characters', () => {
    expect(escape('æˆ‘çˆ±ä½ ')).toBe('u6211u7231u4f60')
    expect(escape('ðŸ˜Š')).toBe('u1f60a')
  })

  it('should escape contiguous mapped symbols', () => {
    expect(escape('[](){}')).toBe('_b_B_p_P_k_K')
  })

  it('should keep default tokens untouched when already escaped', () => {
    const token = '_b'
    expect(escape(token)).toBe(token)
  })

  it('should encode character immediately above ascii boundary', () => {
    const boundaryChar = String.fromCodePoint(MAX_ASCII_CHAR_CODE + 1)
    expect(escape(boundaryChar)).toBe(`u${(MAX_ASCII_CHAR_CODE + 1).toString(16)}`)
  })

  it('should allow custom map entries to override defaults', () => {
    const map = { '[': 'customL', '!': 'customBang' }
    expect(escape('[!', { map })).toBe('customLcustomBang')
  })

  it('should respect custom mapping for leading hyphen', () => {
    const map = { '-': 'dash' }
    expect(escape('-1', { map })).toBe('dash1')
  })

  it('should reuse cached merged map', () => {
    const map = { '[': 'cachedL' }
    expect(escape('[', { map })).toBe('cachedL')
    expect(escape('[', { map })).toBe('cachedL')
  })

  it('should treat empty custom map as default mapping', () => {
    const empty: Record<string, string> = {}
    const source = '![#'
    expect(escape(source, { map: empty })).toBe(escape(source))
  })

  it('should keep default options when none provided', () => {
    expect(escape('abc')).toBe('abc')
  })

  it('should handle Tailwind arbitrary value syntax', () => {
    const source = 'bg-[var(--xx)]'
    const escaped = escape(source)
    expect(escaped).toBe('bg-_bvar_p--xx_P_B')
  })

  it('should handle Tailwind syntax with complex mapping', () => {
    const source = 'bg-[var(--xx)]'
    const escaped = escape(source, { map: ComplexMap })
    expect(escaped).toBe('bg-_cbvar_cp--xx_cP_cB')
  })
})
