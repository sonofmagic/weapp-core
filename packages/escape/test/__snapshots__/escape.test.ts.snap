// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`escape > should match snapshot for complex mapping 1`] = `"_1234567890-_cz_cw_ce_ct_ch_cs_cv_cy_cn_cx_cp_cP__cuqwertyuiop_cb_cB_crQWERTYUIOP_ck_cK_coasdfghjkl_cj_caASDFGHJKL_cc_cqzxcvbnm_cm_cd_cfZXCVBNM_cl_cg_cQ"`;

exports[`escape > should match snapshot for default mapping 1`] = `"_1234567890-_z_w_e_t_h_s_v_y_n_x_p_P__uqwertyuiop_b_B_rQWERTYUIOP_k_K_oasdfghjkl_j_aASDFGHJKL_c_qzxcvbnm_m_d_fZXCVBNM_l_g_Q"`;
